Makefile

```diff
@@ -1,103 +1,187 @@
+	@$(GO_CILINT) run $(GO_CILINT_CHECKERS)
+
+
+#
+# Target for running unit/module tests.
+#
 
 test:
-	@$(GO) test -race -coverprofile=coverage.txt -covermode=atomic $(shell $(GO) list ./... | grep -v vendor)
-
-# clean the given commands
-clean:
-	@for cmd in $(COMMANDS); do \
-	    echo "Cleaning up $$cmd..." && \
-	    cd cmd/$$cmd && \
-	    go clean && \
-	    cd - > /dev/null 2>&1; \
-	done
-
-# build the given images
-images: $(DOCKERDIRS)
-	@for dd in $(DOCKERDIRS); do \
-	    echo "Building docker image for $$dd"; \
-	    buildopts="--quiet"; \
-	    if [ -n "$(DOCKERTAG)" ]; then \
-		buildopts="$$buildopts --tag $(DOCKERTAG)"; \
+	@$(GO_CMD) test -race -coverprofile=coverage.txt -covermode=atomic \
+	    $(GO_MODULES)
+
+# Target for preparing a dist-tarball for release/packaging.
+dist: git-version git-build
+	rm -fr $(TAR_DIR)* && \
+	echo "Creating tarball $(TAR_BALL)..." && \
+	git archive --format=tar --prefix=$(TAR_DIR)/ HEAD > $(TAR_BALL) && \
+	echo "Saving git version and build ids in $(TAR_BALL)..." && \
+	mkdir -p $(TAR_DIR) && cp git-version git-build $(TAR_DIR) && \
+	$(TAR_UPDATE) $(TAR_BALL) $(TAR_DIR)/git-{build,version} && \
+	echo "Compressing $(TAR_BALL) with $(TAR_COMPRESS)..." && \
+	$(TAR_COMPRESS) $(TAR_BALL) && \
+	rm -fr $(TAR_DIR)
+
+git-version:
+	@if [ -d .git ]; then \
+	    version=`git describe 2>/dev/null || echo 0.0.0unknown`; \
+	else \
+	    if [ -f git-version ]; then \
+	        version=`cat git-version`; \
+	    else \
+	        version='0.0.0unknown'; \
 	    fi; \
-	    if [ -n "$(DOCKERPUSH)" ]; then \
-	        buildopts="$$buildopts --registry $(DOCKERPUSH)"; \
+	fi; \
+	echo "$$version" > $@
+
+git-build:
+	@if [ -d .git ]; then \
+	    build=`git rev-parse --short HEAD 2>/dev/null`; \
+	else \
+	    if [ -f git-build ]; then \
+	        build=`cat git-build`; \
+	    else \
+	        build='unknown-build'; \
 	    fi; \
-	    scripts/build/docker-build $$buildopts $$dd; \
-	done
-
-# try to kube-reconfigure and redeploy quasi-random stuff we find lying around
-deploy: images
-	@if [ -n "$(KUBECTL)" ]; then \
-	    for dd in $(DOCKERDIRS); do \
-	        echo "Kube-redeploying $$dd..."; \
-	        for yml in $dd/*config.yaml; do \
-		    echo "Updating configuration $yml..."; \
-		    $(KUBECTL) apply -f $$yml; \
-		done; \
-	        for yml in $dd/*deployment.yaml; do \
-		    echo "Updating deployment $yml..."; \
-		    $(KUBECTL) apply -f $$yml; \
-		done; \
-	    done; \
-	fi
-
-# go-build the given commands
-$(COMMANDS):
-	@echo "Building $@..." && \
-	    cd cmd/$@ && \
-	    GO111MODULE=on go build -o $@
+	fi; \
+	echo "$$build" > $@
 
+# Generic target for recompiling a changed protobuf.
 %.pb.go: %.proto
-	@echo "Generating $@..."
-	@$(PROTOC) -I . $< --go_out=plugins=grpc:.
+	@echo "Generating go code ($@) for updated protobuf $<..."; \
+	$(PROTOC) -I . $< --go_out=plugins=grpc:.
+
+
+# redirect git hooks to our in-repo hook directory
+install-git-hooks:
```

Comment:
Is there intended/planned use for this already? I think it could be mentioned in the commit message, at least.

Reply:
It's for those (like myself) who prefer to have githooks preventing commits which violate certain checks (well go formatting), instead of letting their editors do automatic reformatting before saving. We can mention it in the README.md/#hacking or wherever we want to describe things related to that...

Reply:
Hmm... that's what I had in my last commit (see the PR intro examples about the --version output for a dirty repo). I wonder what the heck I force-pushed the last time.